shelley-ma/formal-spec/value.tex

```diff
@@ -1,77 +1,239 @@
+      & \text{Return the size in words of a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{pointwise} ~\in~ (\Integer \to \Integer \to \Bool) \to (\mathsf{ValType} \to \mathsf{ValType} \to \Bool)\\
+      & \text{Pointwise comparison of two $\mathsf{ValType}$ terms}
+  \end{align*}
+  \caption{The $\ValMonoid$ Structure}
+  \label{fig:ValMonoid}
+\end{figure}
+
+Note that any $\ValMonoid$ must also have the following function, which gives
+the size of the serialized representation of a term of the underlying type.
+This function is used in the UTXO rule.
+We do not include it in the monoid definition because it cannot be defined
+explicity separately from concrete implementation.
+
+\[ \fun{serSize} ~\in~ \mathsf{ValType} \to \MemoryEstimate \]
+
+\subsection{$\Coin$ in the Allegra era}
+
+For the Allegra era, we require the $\ValMonoid$ asset representation to support
+only a single asset, Ada, whose type we call $\Coin$. This section defines the
+$\ValMonoid$ functions for the $\Coin$ type, see Figure
+\ref{fig:coin}. Addition and multiplication are defined in the usual way
+(as on $\Integer$).
 
-When multi-asset support on the ledger is introduced, Ada is still expected to be
+When multi-asset support on the ledger is introduced, Ada ($\Coin$) is still expected to be
 the most common type of token on the ledger.
-The $\Coin$ type is used to represent an amount of Ada.
 It is the only
 type of token that can be used for all non-UTxO ledger accounting, including deposits,
-fees, rewards, treasury, and the proof of stake protocol.
+fees, rewards, treasury, and the proof of stake protocol. For this reason, not
+all occurrences of $\Coin$ inside a transaction or in the ledger state can or
+should be replaced with the abstract $\ValMonoid$.
 
-\begin{figure*}[t!]
-  \emph{Abstract Types}
-  %
+%%
+%% Figure Coin Functions
+%%
+\begin{figure}[htb]
   \begin{align*}
-    \var{aname} \in& ~\AssetName & \text{Asset Names}
+      & \fun{zero} ~=~ 0
+      \nextdef
+      %
+      & \fun{coin} ~=~ \fun{id}_{\Coin}
+      \nextdef
+      %
+      & \fun{inject} ~=~ \fun{id}_{\Coin}
+      \nextdef
+      %
+      & \fun{modifyCoin} ~=~ \fun{id}_{(\Coin \to \Coin)}
+      \nextdef
+      %
+      & \fun{size}~\wcard ~=~ 0
+      \nextdef
+      %
+      & \fun{pointwise} ~=~ \fun{id}_{(\Integer \to \Integer \to \Bool)}
   \end{align*}
-  %
+  \caption{$\ValMonoid$ structure with $\Coin$}
+  \label{fig:coin}
+\end{figure}
+
+
+
+\noindent \textbf{Relation between $\fun{inject}$ and $\fun{coin}$.}
+When composed as
+
+\[\fun{coin} \circ \fun{inject} = \fun{id}_{\Coin}\]
+
+these functions give the identity on $\Coin$, but composed in the opposite order as
+
+\[\fun{inject} \circ \fun{coin}\]
+
+they return a $\ValMonoid$ containing only the Ada tokens contained in the original $\ValMonoid$.
+
+\textbf{Fungible and non-fungible tokens.}
+Recall that a token is said to be fungible at a specific time (slot $s$, for example) if
```

Comment:
I don't think we should mention time here, it has nothing to do with fungibility in our ledger

Reply:
That's how people are making NFTs now - by making them with a policy that is invalid after a given slot. And usually people mean that an NFT is an NFT forever. Whereas we mean it in the weaker sense "no other ones with the same asset ID exist in some ledger state. I think it is useful to disambiguate here

Reply:
I would actually be quite happy with not having this fungibility discussion in the spec at all. Also, I don't think I want to define what a `token` is because it seems to be confusing and not particularly helpful, especially since the definition is not relevant to the ledger at all.